---
title: "18S_DADA2_Jan18_2021"
author: "Swellan Pinto"
date: "18/01/2021"
output: html_document
---

Load packages:
```{r}
library(dada2); packageVersion("dada2")
```

Change to directory containing unzipped trimmed files:
```{r}
path <- "/Users/swellan.pinto/Desktop/18S_Jan13_2021_SLP/trimmed_unzipped_edited" # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path)
```

Read in the names of the fastq files, and perform some string manipulation to get matched lists of the forward and reverse fastq files:
```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.trimmomatic.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.trimmomatic.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

Inspect read quality profiles:
```{r}
plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:10])
```

Filter and trim:
```{r}
# Place filtered files in filtered/subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(227,125),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)
out
```

Learn the error rates:
```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)
```

Sample inference:
```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
dadaFs[[1]]
```

Merge paired reads:
```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE, justConcatenate = TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[5]])
```

Construct sequence table:
```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

Remove chimeras:
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

Track reads through the pipeline:
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
track
```

Assign taxonomy:
```{r}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/swellan.pinto/Desktop/18S_Jan13_2021_SLP/trimmed_unzipped_edited/pr2_version_4.12.0_18S_dada2.fasta.gz", 
  taxLevels = c("Kingdom","Supergroup","Division","Class","Order","Family","Genus","Species"), multithread=TRUE)

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
taxa.print
```

Import into phyloseq:
```{r}
library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
theme_set(theme_bw())
```

```{r}
metadata <- read.csv("/Users/swellan.pinto/Desktop/18S_metadata_Feb05_2021_SLP.csv", header=T, stringsAsFactors = T) 
View(metadata)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), sample_data(metadata), tax_table(taxa))
sample_names(ps)
saveRDS(ps, "/Users/swellan.pinto/Desktop/18S_capuchin_PR2_RL227-125_maxEE2-2.rds")
```
